Marlin/src/sd/disk_io_driver.h

```diff
@@ -0,0 +1,65 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2021 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (c) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
+
+#include <stdint.h>
+
+/**
+ * DiskIO Interace
+ *
+ * Interface for low level disk io
+ */
+class DiskIODriver {
+public:
+  /**
+   * Initialize an SD flash memory card with default clock rate and chip
+   * select pin.  See sd2Card::init(uint8_t sckRateID, uint8_t chipSelectPin).
+   *
+   * \return true for success or false for failure.
+   */
+  virtual bool init(const uint8_t sckRateID, const pin_t chipSelectPin) = 0; //TODO: only for SPI
+
+  virtual bool readStart(const uint32_t block) = 0;
+  virtual bool readData(uint8_t* dst) = 0;
+  virtual bool readStop() = 0;
+
+  virtual bool writeStart(const uint32_t block, const uint32_t) = 0;
+  virtual bool writeData(const uint8_t* src) = 0;
+  virtual bool writeStop() = 0;
+
+  virtual bool readBlock(uint32_t block, uint8_t* dst) = 0;
+  virtual bool writeBlock(uint32_t blockNumber, const uint8_t* src) = 0;
+
+  /**
+   * Read a card's CSD register. The CSD contains Card-Specific Data that
+   * provides information regarding access to the card's contents.
+   *
+   * \param[out] csd pointer to area for returned data.
+   *
+   * \return true for success or false for failure.
+   */
+  virtual bool readCSD(csd_t* csd) = 0;
+
+  virtual uint32_t cardSize() = 0;
+
+  virtual bool isReady() = 0;
```

Comment:
We need to avoid virtual methods at all cost, especially for the AVR / 64K builds. Instead of using virtual methods, instead you should use the Curiously Recursive Template Pattern or follow the methodology employed in the `TMCStepper` library.

Reply:
Even if the virtual is totally abstract, to enforce an interface? 🤔

Reply:
Hum... I don't think CRTP will work here. I'm using runtime polymorphism, not static. So I need a generic base type to hold the pointer for the child, that can changed at runtime. The caller will never know that type of the child, and it can't know, as the user change it at runtime....

The only way I think we can avoid a vtable here, is break the link of `DiskIODriver` with the base classes, and then, do manual casts to the base class: 

Example:
```
changeMedia(reinterpret_cast<DiskIODriver*>(&sd2card_UsbFlashDrive));
```
and break the link with the `DiskIODriver` inteface:
```
class DiskIODriver_SDIO   // DONT INHERIT ANYMORE  : public DiskIODriver {
```

The risk here is that the compiler will not catch errors if the derived types doesn't implement the correct base type methods....

Reply:
Googling around for "alternatives to virtual methods" turns up some interesting articles.